File: NewsDaily/app/src/main/java/com/andruid/magic/newsdaily/paging/NewsViewModel.java
package com.andruid.magic.newsdaily.paging;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.ViewModel;
import androidx.paging.LivePagedListBuilder;
import androidx.paging.PagedList;

import com.andruid.magic.newsloader.model.News;

public class NewsViewModel extends ViewModel {
    private LiveData<PagedList<News>> pagedListLiveData;

    public NewsViewModel() {
        NewsDataSourceFactory dataSourceFactory = new NewsDataSourceFactory("in");
        PagedList.Config pagedListConfig = new PagedList.Config.Builder()
                .setEnablePlaceholders(false)
                .setPageSize(NewsDataSource.PAGE_SIZE)
                .build();
        pagedListLiveData = new LivePagedListBuilder<>(dataSourceFactory, pagedListConfig).build();
    }

    public LiveData<PagedList<News>> getPagedListLiveData() {
        return pagedListLiveData;
    }
}

File: NewsDaily/app/src/main/java/com/death14stroke/newsdaily/ui/viewmodel/NewsViewModel.kt
package com.death14stroke.newsdaily.ui.viewmodel

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import androidx.paging.Pager
import androidx.paging.PagingConfig
import androidx.paging.cachedIn
import com.death14stroke.newsdaily.data.PAGE_SIZE
import com.death14stroke.newsdaily.data.model.Result
import com.death14stroke.newsdaily.data.repository.MainRepository
import com.death14stroke.newsdaily.paging.NewsPagingSource
import com.death14stroke.newsloader.data.model.Category
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.flow.*

class NewsViewModel(private val repository: MainRepository, category: Category) : ViewModel() {
    private val config = PagingConfig(PAGE_SIZE)
    val countryFlow = MutableStateFlow(repository.getSelectedCountry())

    @OptIn(ExperimentalCoroutinesApi::class)
    val newsFlow = countryFlow.flatMapLatest { country ->
        Pager(config) { NewsPagingSource(repository, country, category) }
            .flow
            .cachedIn(viewModelScope)
            .map { news -> Result.Success(news) }
            .flowOn(Dispatchers.IO)
    }.stateIn(viewModelScope, SharingStarted.Lazily, Result.Loading)
}